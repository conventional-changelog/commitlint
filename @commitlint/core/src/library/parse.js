import {sync} from '@marionebl/conventional-commits-parser';
import defaultChangelogOpts from 'conventional-changelog-angular';

export default parse;

async function parse(message, parser = sync, parserOpts) {
	if (!parserOpts) {
		const changelogOpts = await defaultChangelogOpts;
		parserOpts = changelogOpts.parserOpts;
	}

	const parsed = parser(message, parserOpts);
	parsed.raw = message;
	return parsed;
}
