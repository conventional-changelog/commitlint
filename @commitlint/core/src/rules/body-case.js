import ensureCase from '../library/ensure-case';
import message from '../library/message';

export default (parsed, when, value) => {
	const {body} = parsed;

	if (!body) {
		return [true];
	}

	const negated = when === 'never';

	const result = ensureCase(body, value);
	return [
		negated ? !result : result,
		message([`body must`, negated ? `not` : null, `be ${value}`])
	];
};
