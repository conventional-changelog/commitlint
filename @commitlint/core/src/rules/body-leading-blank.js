import toLines from '../library/to-lines';

export default (parsed, when) => {
	// Flunk if no body is found
	if (!parsed.body) {
		return [true];
	}

	const negated = when === 'never';
	const [leading] = toLines(parsed.raw).slice(1);

	// Check if the first line of body is empty
	const succeeds = leading === '';

	return [
		negated ? !succeeds : succeeds,
		['body', negated ? 'may not' : 'must', 'have leading blank line']
			.filter(Boolean)
			.join(' ')
	];
};
