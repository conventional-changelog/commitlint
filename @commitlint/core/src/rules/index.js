export default {
	'body-case': require('./body-case'),
	'body-empty': require('./body-empty'),
	'body-leading-blank': require('./body-leading-blank'),
	'body-max-length': require('./body-max-length'),
	'body-min-length': require('./body-min-length'),
	'body-tense': require('./body-tense'),
	'footer-empty': require('./footer-empty'),
	'footer-leading-blank': require('./footer-leading-blank'),
	'footer-max-length': require('./footer-max-length'),
	'footer-min-length': require('./footer-min-length'),
	'footer-tense': require('./footer-tense'),
	'header-max-length': require('./header-max-length'),
	'header-min-length': require('./header-min-length'),
	lang: require('./lang'),
	'scope-case': require('./scope-case'),
	'scope-empty': require('./scope-empty'),
	'scope-enum': require('./scope-enum'),
	'scope-max-length': require('./scope-max-length'),
	'scope-min-length': require('./scope-min-length'),
	'signed-off-by': require('./signed-off-by'),
	'subject-case': require('./subject-case'),
	'subject-empty': require('./subject-empty'),
	'subject-full-stop': require('./subject-full-stop'),
	'subject-leading-capital': require('./subject-leading-capital'),
	'subject-max-length': require('./subject-max-length'),
	'subject-min-length': require('./subject-min-length'),
	'subject-tense': require('./subject-tense'),
	'type-case': require('./type-case'),
	'type-empty': require('./type-empty'),
	'type-enum': require('./type-enum'),
	'type-max-length': require('./type-max-length'),
	'type-min-length': require('./type-min-length')
};
