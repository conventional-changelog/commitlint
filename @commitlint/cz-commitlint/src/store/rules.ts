import {QualifiedRules} from '@commitlint/types';
import {Rule} from '../types';

const storeKey = Symbol('rules');

const store: {
	[storeKey]: QualifiedRules;
} = {
	[storeKey]: {},
};

export function getRule(key: string, property: string): Rule | undefined {
	if (key.split('-').length > 1) {
		return;
	}
	return store[storeKey][`${key}-${property}`];
}

export function setRules(newRules: QualifiedRules): void {
	store[storeKey] = newRules;
}

export type GetRuleMethod = typeof getRule;
export type SetRulesMethod = typeof setRules;
