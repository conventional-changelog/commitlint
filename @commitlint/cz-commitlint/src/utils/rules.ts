import {RuleConfigSeverity} from '@commitlint/types';
import {Rule} from '../types';

export function ruleIsDisabled(
	rule: Rule
): rule is Readonly<[RuleConfigSeverity.Disabled]> {
	if (rule && Array.isArray(rule) && rule[0] === RuleConfigSeverity.Disabled) {
		return true;
	}
	return false;
}

/**
 * Check if a rule definition is active
 * @param rule to check
 * @return if the rule definition is active
 */
export function ruleIsActive<T extends Rule>(
	rule: T
): rule is Exclude<T, Readonly<[RuleConfigSeverity.Disabled]>> {
	if (rule && Array.isArray(rule)) {
		return rule[0] > RuleConfigSeverity.Disabled;
	}
	return false;
}

/**
 * Check if a rule definition is applicable
 * @param rule to check
 * @return if the rule definition is applicable
 */
export function ruleIsApplicable(
	rule: Rule
): rule is
	| Readonly<[RuleConfigSeverity, 'always']>
	| Readonly<[RuleConfigSeverity, 'always', unknown]> {
	if (rule && Array.isArray(rule)) {
		return rule[1] === 'always';
	}
	return false;
}

/**
 * Check if a rule definition is applicable
 * @param rule to check
 * @return if the rule definition is applicable
 */
export function ruleIsNotApplicable(
	rule: Rule
): rule is
	| Readonly<[RuleConfigSeverity, 'never']>
	| Readonly<[RuleConfigSeverity, 'never', unknown]> {
	if (rule && Array.isArray(rule)) {
		return rule[1] === 'never';
	}
	return false;
}

export function enumRuleIsActive(
	rule: Rule
): rule is Readonly<
	[RuleConfigSeverity.Warning | RuleConfigSeverity.Error, 'always', string[]]
> {
	return (
		ruleIsActive(rule) &&
		ruleIsApplicable(rule) &&
		Array.isArray(rule[2]) &&
		rule[2].length > 0
	);
}

export function getEnumList(rule: Rule): string[] {
	return Array.isArray(rule[2]) ? rule[2] : [];
}

export function getMaxLength(rule?: Rule): number {
	if (
		rule &&
		ruleIsActive(rule) &&
		ruleIsApplicable(rule) &&
		typeof rule[2] === 'number'
	) {
		return rule[2];
	}
	return Infinity;
}

export function getMinLength(rule?: Rule): number {
	if (
		rule &&
		ruleIsActive(rule) &&
		ruleIsApplicable(rule) &&
		typeof rule[2] === 'number'
	) {
		return rule[2];
	}
	return 0;
}
