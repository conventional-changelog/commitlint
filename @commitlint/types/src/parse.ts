import type {Commit, Options} from 'conventional-commits-parser';

export type Parser = (message: string, options: Options) => Omit<Commit, 'raw'>;
