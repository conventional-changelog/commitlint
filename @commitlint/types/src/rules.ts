import {TargetCaseType} from './ensure';
import {Commit} from './parse';

/**
 * Rules match the input either as successful or failed.
 * For example, when `header-full-stop` detects a full stop and is set as "always"; it's true.
 * If the `header-full-stop` discovers a full stop but is set to "never"; it's false.
 */
export type RuleOutcome = Readonly<[boolean, string?]>;

/**
 * Rules receive a parsed commit, condition, and possible additional settings through value.
 * All rules should provide the most sensible rule condition and value.
 */
export type RuleType = 'async' | 'sync' | 'either';

export type BaseRule<Value = never, Type extends RuleType = 'either'> = (
	parsed: Commit,
	when?: RuleConfigCondition,
	value?: Value
) => Type extends 'either'
	? RuleOutcome | Promise<RuleOutcome>
	: Type extends 'async'
	? Promise<RuleOutcome>
	: Type extends 'sync'
	? RuleOutcome
	: never;

export type Rule<Value = never> = BaseRule<Value, 'either'>;
export type AsyncRule<Value = never> = BaseRule<Value, 'async'>;
export type SyncRule<Value = never> = BaseRule<Value, 'sync'>;

/**
 * Rules always have a severity.
 * Severity indicates what to do if the rule is found to be broken
 * 0 - Disable this rule
 * 1 - Warn for violations
 * 2 - Error for violations
 */
export enum RuleConfigSeverity {
	Disabled = 0,
	Warning = 1,
	Error = 2,
}

/**
 * Rules always have a condition.
 * It can be either "always" (as tested), or "never" (as tested).
 * For example, `header-full-stop` can be enforced as "always" or "never".
 */
export type RuleConfigCondition = 'always' | 'never';

export type RuleConfigTuple<T> = T extends void
	?
			| Readonly<[RuleConfigSeverity.Disabled]>
			| Readonly<[RuleConfigSeverity, RuleConfigCondition]>
	:
			| Readonly<[RuleConfigSeverity.Disabled]>
			| Readonly<[RuleConfigSeverity, RuleConfigCondition, T]>;

export enum RuleConfigQuality {
	User,
	Qualified,
}

export type QualifiedRuleConfig<T> =
	| (() => RuleConfigTuple<T>)
	| (() => Promise<RuleConfigTuple<T>>)
	| RuleConfigTuple<T>;

export type RuleConfig<
	V = RuleConfigQuality.Qualified,
	T = void
> = V extends RuleConfigQuality.Qualified
	? RuleConfigTuple<T>
	: QualifiedRuleConfig<T>;

export type CaseRuleConfig<V = RuleConfigQuality.User> = RuleConfig<
	V,
	TargetCaseType | TargetCaseType[]
>;
export type LengthRuleConfig<V = RuleConfigQuality.User> = RuleConfig<
	V,
	number
>;
export type EnumRuleConfig<V = RuleConfigQuality.User> = RuleConfig<
	V,
	string[]
>;

export type RulesConfig<V = RuleConfigQuality.User> = {
	'body-case': CaseRuleConfig<V>;
	'body-empty': RuleConfig<V>;
	'body-full-stop': RuleConfig<V, string>;
	'body-leading-blank': RuleConfig<V>;
	'body-max-length': LengthRuleConfig<V>;
	'body-max-line-length': LengthRuleConfig<V>;
	'body-min-length': LengthRuleConfig<V>;
	'footer-empty': RuleConfig<V>;
	'footer-leading-blank': RuleConfig<V>;
	'footer-max-length': LengthRuleConfig<V>;
	'footer-max-line-length': LengthRuleConfig<V>;
	'footer-min-length': LengthRuleConfig<V>;
	'header-case': CaseRuleConfig<V>;
	'header-full-stop': RuleConfig<V, string>;
	'header-max-length': LengthRuleConfig<V>;
	'header-min-length': LengthRuleConfig<V>;
	'references-empty': RuleConfig<V>;
	'scope-case': CaseRuleConfig<V>;
	'scope-empty': RuleConfig<V>;
	'scope-enum': EnumRuleConfig<V>;
	'scope-max-length': LengthRuleConfig<V>;
	'scope-min-length': LengthRuleConfig<V>;
	'signed-off-by': RuleConfig<V, string>;
	'subject-case': CaseRuleConfig<V>;
	'subject-empty': RuleConfig<V>;
	'subject-full-stop': RuleConfig<V, string>;
	'subject-max-length': LengthRuleConfig<V>;
	'subject-min-length': LengthRuleConfig<V>;
	'trailer-exists': RuleConfig<V, string>;
	'type-case': CaseRuleConfig<V>;
	'type-empty': RuleConfig<V>;
	'type-enum': EnumRuleConfig<V>;
	'type-max-length': LengthRuleConfig<V>;
	'type-min-length': LengthRuleConfig<V>;
	// Plugins may add their custom rules
	[key: string]: AnyRuleConfig<V>;
};

export type AnyRuleConfig<V> = RuleConfig<V, unknown> | RuleConfig<V, void>;
