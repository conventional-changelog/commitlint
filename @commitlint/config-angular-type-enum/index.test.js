import config from '.';
import rules from '@commitlint/rules';
import parse from '@commitlint/parse';

const messages = Object.entries({
	simple: 'test: subject',
	empty: 'test: subject\nbody',
	filled: 'test: subject\nBREAKING CHANGE: something important'
});
const configRules = Object.entries(config.rules);

it('should have correct structure', () => {
	expect(config).toMatchObject({
		rules: expect.any(Object)
	});
});

describe('rules', () => {
	describe.each(configRules)('%s', (name, options) => {
		const severity = options.unshift();
		it.each(messages)('%s', async (type, message) => {
			expect(
				rules[name](await parse(message), severity, options)
			).toMatchSnapshot();
		});
	});
});
