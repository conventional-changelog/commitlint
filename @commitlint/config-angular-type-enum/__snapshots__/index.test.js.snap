// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules type-enum empty 1`] = `
Array [
  true,
  "type must be one of [build, ci, docs, feat, fix, perf, refactor, revert, style, test]",
]
`;

exports[`rules type-enum filled 1`] = `
Array [
  true,
  "type must be one of [build, ci, docs, feat, fix, perf, refactor, revert, style, test]",
]
`;

exports[`rules type-enum simple 1`] = `
Array [
  true,
  "type must be one of [build, ci, docs, feat, fix, perf, refactor, revert, style, test]",
]
`;
